File: RealizeQueue/Queue/Queue.cs
namespace Queue
{
    using System.Collections.Generic;

    public class Queue<T>
    {
        public Queue()
        {
            list = new List<QueueElement<T>>();
        }

        /// <summary>
        /// Add element in the end of Queue.
        /// </summary>
        /// <param name="value"> Added element. </param>
        /// <param name="priority"> Priority of element. </param>
        public void Enqueue(T value, int priority)
        {
            QueueElement<T> newElement;
            newElement.priority = priority;
            newElement.value = value;
            list.Add(newElement);
        }

        /// <summary>
        /// Return element with the highest priority.
        /// </summary>
        /// <returns></returns>
        public object Dequeue()
        {
            if (list.Count == 0)
                throw new EmptyQueueException();

            int IndexMaxPriority = 0;
            T result = default(T);

            for (int i = 0; i < list.Count; i++)
                if (list[i].priority > list[IndexMaxPriority].priority)
                    IndexMaxPriority = i;

            result = list[IndexMaxPriority].value;
            QueueElement<T> temp;
            temp.value = result;
            temp.priority = IndexMaxPriority;
            list.Remove(temp);

            return result;
        }

        /// <summary>
        /// Test queue for empty.
        /// </summary>
        /// <returns></returns>
        public bool TestForEmpty()
        {
            return list.Count == 0;
        }

        /// <summary>
        /// Element of Queue.
        /// </summary>
        /// <typeparam name="T"></typeparam>
        private struct QueueElement<T>
        {
            public T value;
            public int priority;
        }

        private List<QueueElement<T>> list;
    }
}

File: 4sem/task3/task3/Queue.fs
exception EmptyQueueException of string

let rec remove index list =
    match index, list with
    | 0, x::xs -> xs
    | index, x::xs -> x::remove (index - 1) xs
    | index, [] -> failwith "index out of range"
        
type 'a QueueElement =
       struct        
          val Value: 'a
          val Priority: int
          new(value: 'a, priority: int) = { Value = value; Priority = priority }
       end

type 'a Queue() =
    let mutable list : 'QueueElement list = []

    member this.isEmpty = list.Length = 0

    member this.enqueue(value, priority) = 
        let element = new QueueElement<'a> (value, priority)
        list <- list @ [element]

    member this.dequeue =
        if (list.IsEmpty) then
            raise (EmptyQueueException "failed dequeue method. The Queue is empty!")

        let mutable indexMaxPriority = 0;
        for i = 0 to list.Length - 1 do
            if (list.Item(i).Priority > list.Item(indexMaxPriority).Priority) then
                indexMaxPriority <- i       
        let result = list.Item(indexMaxPriority).Value
        list <- remove indexMaxPriority list
        result

let testQueue = Queue<int>()

testQueue.enqueue (2, 2)
testQueue.enqueue (0, 7)
testQueue.enqueue (8, 5)

printfn "%A" testQueue.dequeue
printfn "%A" testQueue.dequeue
printfn "%A" testQueue.dequeue

try
    printfn "%A" testQueue.dequeue
with
    | EmptyQueueException str -> printfn "%s" str

printfn "%A" testQueue.isEmpty